kernel/src/drivers/ps2/io.rs

```diff
@@ -6,127 +6,148 @@ static COMMAND_PORT: IOPort = unsafe { IOPort::new(0x64) };
+    command_raw(cmd as u8)?;
+    write(&DATA_PORT, data)?;
+
+    Ok(())
 }
 
 /// Sends a controller command with a return
-pub fn command_ret(cmd: ControllerReturnCommand) -> Option<u8> {
-    command_raw(cmd as u8);
+pub fn command_ret(cmd: ControllerReturnCommand) -> Result<u8, Ps2Error> {
+    command_raw(cmd as u8)?;
     read(&DATA_PORT)
 }
 
 /// Sends a raw controller command code
-fn command_raw(cmd: u8) {
-    write(&COMMAND_PORT, cmd);
+fn command_raw(cmd: u8) -> Result<(), Ps2Error> {
+    write(&COMMAND_PORT, cmd)
 }
 
 /// Writes the given value to the data port
-pub fn write_data(value: u8) {
-    write(&DATA_PORT, value);
+pub fn write_data(value: u8) -> Result<(), Ps2Error> {
+    write(&DATA_PORT, value)
 }
 
 /// Reads from the data port
-pub fn read_data() -> Option<u8> {
+pub fn read_data() -> Result<u8, Ps2Error> {
     read(&DATA_PORT)
 }
 
 /// Writes to the given port, or waits until available
-pub fn write(port: &IOPort, value: u8) {
-    wait_write();
+pub fn write(port: &IOPort, value: u8) -> Result<(), Ps2Error> {
+    wait_write()?;
     port.write(value);
+
+    Ok(())
 }
 
 /// Reads from the given port, returning an optional value. None returned if nothing could be read
-pub fn read(port: &IOPort) -> Option<u8> {
-    if wait_read() {
-        Some(port.read())
-    } else {
-        None
-    }
+pub fn read(port: &IOPort) -> Result<u8, Ps2Error> {
+    wait_read()?;
+    Ok(port.read())
 }
 
 /// Flushes the controller's output buffer
-pub fn flush_output() {
-    loop {
-        // Read until the output status bit is empty
-        if check_status(OUTPUT_STATUS_BIT) {
-            DATA_PORT.read();
-        } else {
-            break;
-        }
+pub fn flush_output() -> Result<(), Ps2Error> {
+    // Read until the output status bit is empty
+    while can_read()? {
+        DATA_PORT.read();
     }
+    Ok(())
 }
 
-/// Waits for the write status bit to empty
-fn wait_write() {
+/// Reads from the status port and returns the flags
+pub fn read_status() -> Result<StatusFlags, Ps2Error> {
+    Ok(StatusFlags::from_bits_truncate(STATUS_PORT.read()))
+}
+
+/// Returns true if the write status bit is 0
+pub fn can_write() -> Result<bool, Ps2Error> {
+    read_status().map(|status| !status.contains(StatusFlags::INPUT_FULL))
+}
+
+/// Returns true if the read status bit is 1
+pub fn can_read() -> Result<bool, Ps2Error> {
+    read_status().map(|status| status.contains(StatusFlags::OUTPUT_FULL))
+}
+
+/// Waits for the write status bit to equal 0
+fn wait_write() -> Result<(), Ps2Error> {
     loop {
         // Check if the input status bit is empty
-        if !check_status(INPUT_STATUS_BIT) {
-            break;
+        if can_write()? {
+            break
         }
     }
+
+    Ok(())
 }
 
 /// Waits for the read status bit to equal 1, and returns true if successful
-fn wait_read() -> bool {
+fn wait_read() -> Result<(), Ps2Error> {
```

Comment:
I think this function name is slightly confusing -- it sounds like it waits and then reads.

Reply:
This is only used once, too... This could be removed probably. We should look for similar cases where a function is only used once

Reply:
Hmm, yes. These should be inlined.